ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -115,10 +128,29 @@ private ChannelInboundHandler getServerHandler(){
       public void channelRead(ChannelHandlerContext ctx, Object msg) {
         final DataStreamRequestByteBuf request = (DataStreamRequestByteBuf)msg;
         final ByteBuf buf = request.slice();
-        final AtomicBoolean released = new AtomicBoolean();
-        streams.computeIfAbsent(request.getStreamId(), id -> getDataStreamFuture(buf, released))
-            .thenApply(stream -> writeTo(buf, stream, released.get()))
-            .thenAccept(byteWritten -> sendReply(request, byteWritten, ctx));
+        final boolean isHeader = request.getStreamOffset() == -1;
+
+        CompletableFuture<?>[] parallelWrites = new CompletableFuture<?>[clients.size() + 1];
+
+        final CompletableFuture<?> localWrites = isHeader?
+                streams.computeIfAbsent(request.getStreamId(), id -> getDataStreamFuture(buf))
+                : streams.get(request.getStreamId()).thenApply(stream -> writeTo(buf, stream));
+        parallelWrites[0] = localWrites;
+        peersStreamOutput.putIfAbsent(request.getStreamId(), getDataStreamOutput());
+
+          // do not need to forward header request
+        if (isHeader) {
+          for (int i = 0; i < peersStreamOutput.get(request.getStreamId()).size(); i++) {
+            parallelWrites[i + 1] = peersStreamOutput.get(request.getStreamId()).get(i).getHeaderFuture();
+          }
```

Comment:
Do not understand why add header future 2 times, when there are 2 clients.

Reply:
The header happens when we create a DataStreamOutput: https://github.com/apache/incubator-ratis/blob/88327daff3261cb3e2929d7fbcbba1ec51d67a34/ratis-client/src/main/java/org/apache/ratis/client/impl/DataStreamClientImpl.java#L79

so for the first header request, it will trigger two other header requests when we initialize two DataStreamOutput for current stream. Then we need to wait for all header requests succeed before reply to client to say its header request succeed.

Reply:
`peersStreamOutput.putIfAbsent(request.getStreamId(), getDataStreamOutput());` will trigger other header requests when receive a header request (i.e. a new stream id).